pkg/network/server.go

```diff
@@ -579,26 +603,52 @@ func (s *Server) IsInSync() bool {
 	var peersNumber int
 	var notHigher int
 
-	if s.MinPeers == 0 {
+	ourLastBlock := s.chain.BlockHeight()
+	if !s.canStartSync.Load() { // TODO: should be removed, `canStartSync` should be updated during BlockCMD/MPTDataCMD processing
+		p := s.p.Load()
+		if p != 0 && ourLastBlock >= p && s.mptPool.Count() == 0 {
+			s.canStartSync.CAS(false, true)
+		}
+	}
+	if s.MinPeers == 0 && s.canStartSync.Load() {
 		return true
 	}
 
-	ourLastBlock := s.chain.BlockHeight()
-
 	s.lock.RLock()
+	heights := make([]uint32, 0)
 	for p := range s.peers {
 		if p.Handshaked() {
 			peersNumber++
-			if ourLastBlock >= p.LastBlockIndex() {
+			peerLastBlock := p.LastBlockIndex()
+			if ourLastBlock >= peerLastBlock {
 				notHigher++
 			}
+			i := sort.Search(len(heights), func(i int) bool {
+				return heights[i] >= peerLastBlock
+			})
+			heights = append(heights, peerLastBlock)
+			if i != len(heights)-1 {
+				copy(heights[i+1:], heights[i:])
+				heights[i] = peerLastBlock
+			}
 		}
 	}
 	s.lock.RUnlock()
 
+	if !s.canStartSync.Load() && s.p.Load() == 0 && peersNumber >= s.MinPeers && len(heights) > 0 {
+		// choose the height of the median peer as current chain's height
+		p1 := (heights[len(heights)/2] / uint32(s.chain.GetConfig().StateSyncInterval)) * uint32(s.chain.GetConfig().StateSyncInterval)
+		if p1 < uint32(2*s.chain.GetConfig().StateSyncInterval) {
+			// chain is too low to start state exchange process, use the standard sync mechanism
+			s.canStartSync.CAS(false, true)
+		} else if s.p.CAS(0, p1) {
+			s.log.Info("try to sync state for latest state synchronization point", zap.Uint32("height", p1))
```

Comment:
`p1` needs to be saved in the DB, that's what we're working against.

Reply:
I thought that we may go without storing this `p1` value, but it's more reliable to store it, so fixed.